File: src/csharp/KernelEngine.Kernel/ConsoleSink.cs
using KernelEngine.Kernel.Native;

namespace KernelEngine.Kernel;

/// <summary>
/// Writes log events to <see cref="Console.Error"/> using the same
/// <c>[LEVEL] tag: message</c> format as the native <c>ke_console_sink</c>.
/// Level filtering is handled by the logger before this sink is called.
/// </summary>
public sealed class ConsoleSink : ILoggerSink
{
    private static readonly string[] s_levels = ["TRACE", "DEBUG", "INFO", "WARN", "ERROR", "CRIT"];

    /// <inheritdoc/>
    public ke_log_level MinLevel { get; init; } = ke_log_level.KE_LOG_LEVEL_TRACE;

    /// <inheritdoc/>
    public void Log(ke_log_level level, string tag, string message)
    {
        int idx = (int)level;
        string label = (idx >= 0 && idx < s_levels.Length) ? s_levels[idx] : "?";
        Console.Error.WriteLine($"[{label}] {tag}: {message}");
        Console.Error.Flush();
    }
}

File: src/c/kernel/src/logger/console_sink.c
#include <kernel_engine/kernel/logger/console_sink.h>
#include <stdio.h>

static void console_sink_log(ke_logger_sink *self, const ke_log_event *event)
{
    (void)self;
    fprintf(stderr, "[%s] %s: %s\n",
            ke_log_level_to_string(event->level),
            event->tag     ? event->tag     : "",
            event->message ? event->message : "");
    fflush(stderr);
}

ke_logger_sink ke_console_sink_create(ke_log_level min_level)
{
    ke_logger_sink sink;
    sink.handle    = NULL;
    sink.min_level = (int32_t)min_level;
    sink.log       = console_sink_log;
    sink.destroy   = NULL;
    return sink;
}
